sql/pg_net--0.1.sql

```diff
@@ -8,6 +8,7 @@ create table net.http_request_queue(
     params jsonb not null,
     headers jsonb not null,
     timeout_milliseconds int not null,
+    error_msg text,
```

Comment:
Needed for specifying the request error message.

Reply:
I think `error_msg` wants to be on `net.http_response` rather than the request queue. 

Currently, all errors, including local ones like `timed_out`, are persisted in `net.http_response`. If we add an `error_msg`to the request queue we'll have to check in the `http_request_queue` table in the `collect_response` function any time there is no value in `http_response`. We also won't be able to expose the errors that we find because the return type is `http_response` which has no field for `error_msg`

In other words, there is no public API to expose error messages in the request queue

Reply:
Yeah, you're right. I'll change the error message to be on the response table.

Reply:
> I think error_msg wants to be on net.http_response rather than the request queue.

@olirice On second thought, I don't think so. This is a **request error**, if I were to insert this on the response table that would mean it has to have a `status_code` and actually it doesn't(it's NULL) when there's a request error. 

There's probably more stuff that has to be changed from the `http_request_queue`. For starters the `timeout_milliseconds` should be a global config setting, not a column, the same goes for `delete_after`. Removing the `queue` from the name would also be good I think(queue is the use we give to the table, but otherwise is a regular entity).

Anyway, that's a lot of changes. I think they should be done on another PR. As it is, this PR is useful because it makes the extension more robust.

Reply:
`timed_out bool` could also be on the `http_request_queue` as it's also a property of the request - there is **no response** yet.

Reply:
>  If we add an error_msgto the request queue we'll have to check in the http_request_queue table in the collect_response

Hm, maybe the `collect_response` interface can change as well. If you notice, right now the worker does a LEFT JOIN, ideally it'd just do a SELECT over the request queue, less work inside would be better for performance/robustness.

Reply:
@olirice Right now I do this query to avoid repeating invalid requests:

```sql
SELECT
q.id, q.url, q.error_msg
FROM net.http_request_queue q\
LEFT JOIN net.http_response r ON q.id = r.id
WHERE r.id IS NULL AND q.error_msg is NULL
-- when the response is null and the request has no error message(valid)
```

How would I avoid processing invalid requests when changing the `error_msg` to response.table?

> I think error_msg wants to be on net.http_response rather than the request queue.

Actually it seems to me that `error_msg` wants to be on the request queue - since it's a **request error** and not a response(has no status code nor headers nor body).

Reply:
> How would I avoid processing invalid requests when changing the error_msg to response.table?

If `error_message` was on the response, we'd then have a response record on failure so the query would be
```sql
select
   q.id, q.url
from
   net.http_request_queue q
   left join net.http_response r
      on q.id = r.id
where
    r.id is null
```

> Actually it seems to me that error_msg wants to be on the request queue - since it's a request error and not a response(has no status code nor headers nor body)

I see where you're coming from. Since request to response has a 1:1 relationship, and both tables are internal, I'm not too concerned about what the structure in the DB looks like. What's important to figure out is how to expose client/request errors & server/http side errors to users coherently through the public API

The public method for retrieving a response is `net.collect_response`, which has the return type `net.http_resposne`. Since that record would not include client/request errors, we don't have a way to let the user know when a client/request error has occurred.

In a perfect world we'd wrap the http response in [result type](https://doc.rust-lang.org/std/result/) and that's what would be returned from `net.collect_response`

```sql
-----------
-- TYPES --
-----------

-- A response from an HTTP server (abbreviated)
create type http_response as (request_id bigint, http_status_code int, contents text);

-- Potential states of an HTTP Request
create type request_status as enum ('PENDING', 'SUCCESS', 'ERROR');

-- State wrapper around (any protocol) responses
create type net_result as (status request_status, response http_response, msg text);

--------------
-- Examples --
--------------

-- A (non-http) error has occured e.g. timeout
select ('ERROR', null, 'client side timeout occurred')::net_result;
select ('ERROR', null, 'invalid protocol')::net_result;

-- Has not been processed
select ('PENDING', null, 'request has not been processed')::net_result;

-- Successful request resulting in http 404
select ('SUCCESS', (1, 404, 'Not Found')::http_response, 'ok')::net_result;

-- Successful request resulting in http 200
select ('SUCCESS', (1, 200, 'Some data')::http_response, 'ok')::net_result;
```

So in those examples, `net.collect_response` would return a `net_result` (`net.result`). I avoided this previously because ORMs don't always have good support for enums and composite types. Would postgrest be able to handle that response type ok @steve-chavez?

If so, that wrapping approach feels like it will be the most flexible for us going forward. What do you think?

Reply:
> If error_message was on the response, we'd then have a response record on failure so the query would be

Cool. I've changed the error message to be on the response table.

>  I avoided this previously because ORMs don't always have good support for enums and composite types. Would postgrest be able to handle that response type ok @steve-chavez?

Yeah, it should be fine. PostgREST doesn't use an ORM, if the types can be represented as json it should be fine. Though I don't know if this extension or its methods will be used through PostgREST, they seem more internal to me.